website documentation for timelock (#173)
drand/website: docs/docs/timelock-encryption.md

    $U = rG_2\in \mathbb{G_2}$, "the ephemeral public key"

    $V = \theta \oplus H_2(rG_{id})$, "the mask commitment"

    $W = M \oplus H_4(\theta)$, "the one-time pad"

### Decryption

A client that wishes to decrypt a ciphertext $C$ takes the associated signature of epoch $p: \pi_p \in \mathbb{G_1}$ and performs the following:

1. Compute $\theta = V \oplus H_2(e(\pi_p, U))$
2. Compute $M = W \oplus H_4(\theta)$
3. Set $r = H_3(\theta, M)$
4. Test that $U = rG_1 $ if not, reject
5. $M$ is the corresponding plaintext

## Libraries available for timelock encryption

Listed below are various libraries and clients for using timelock encryption. Details on how to install and use them are available at the links provided.

- [tlock](https://github.com/drand/tlock)

tlock is a go CLI and library developed by [Ardan Labs](https://www.ardanlabs.com/) and maintained by the drand team for performing timelock encryption and decryption

- [tlock-js](https://github.com/drand/tlock-js)

tlock-js is a javascript library developed and maintained by the drand team for performing timelock encryption and decryption written in Typescript

- [dee](https://github.com/thibmeu/drand-rs)

> dee is a rust CLI developed and maintained by Thibault Meunier from cloudflare. It can perform timelock encryption and decryption, as well as retrieve beacons from any of the drand networks.

Reviewer: ```suggestion
dee is a rust CLI developed and maintained by Thibault Meunier from Cloudflare. It can perform timelock encryption and decryption, as well as retrieve beacons from any of the drand networks.
```

- **neither the scheme, nor drand are quantum resistant**

The cryptography we employ do not use quantum resistant algorithms. If you timelock encrypt something for 1000 years and a viable quantum computer becomes available, anybody can decrypt it. Right now there are no widespread quantum resistant schemes for threshold cryptography, but we're keeping an eye out for the future!
To that end, consider it relatively safe to timelock encrypt something for 5-20 years depending on how bullish you are on quantum computing.

- **if the League of Entropy shuts down, members will delete their keys**

The world is unpredictable (just like drand randomness... heh), and it's possible that the League of Entropy will all hang up their coats at some point in the future. Should that happen, members would have two choices with their private keys: release them to the world or delete them entirely.
The former would mean that ciphertexts created for some time after the cessation of the network would be decryptable to everybody. The latter would mean that ciphertexts created for some time after the cessation of the network would be unencryptable forever (/until quantum computers can break them).

> In the interests of privacy, we felt the latter option was preferable. That said, if you encrypt the private key to your Filecoin fortune to stop yourself from spending it now and the network stops... you're going to have a bad time.

Reviewer: ```suggestion
In the interests of privacy, we felt the latter option was preferable. That said, if you encrypt the private key to your [insert cryptocurrency name] fortune to stop yourself from spending it now and the network stops... you're going to have a bad time.
```

Author: ffs I can't believe you're removing this top shill opportunity :p

> [1] their report is available in timelocked form [here](https://research.kudelskisecurity.com/2023/01/09/announcing-a-timelocked-responsible-disclosure/), although it was encrypted using the drand testnet, so you will need to use the [tlock go library](https://github.com/drand/tlock) to decrypt it by passing in the testnet details. All the security issues they identified have since been remediated! You can find remediation details in the github issues of the relevant repos.

Reviewer: It's not. Actually that's just the "date" issue, not the whole report.

Author: Ah hmm, did they timelock the whole report anywhere?!

Reviewer: Not that I know of, and we are yet to do a public blog post and release of it I think. Let's coordinate with them

Author: Grand, I'll modify this section to say that they just timelocked part of it

## Use cases

We have identified some cool use cases for timelock encryption, but these are certainly not exhaustive - hopefully you can come up with new, innovative use cases we haven't thought of!

- [responsible vulnerability disclosure](https://timevault.drand.love/)
- transfer of assets/password upon death
- sealed bid auctions
- [MEV prevention](https://coinmarketcap.com/alexandria/glossary/miner-extractable-value-mev)
- public turn-based games
- voting
- quizzes

> 

Reviewer: Maybe link the [#drand-use-cases](https://drandworkspace.slack.com/archives/C04UANC4P7D) channel from slack?

Author: have added some info here

## The math bit 🤓

### Encryption

A client that wishes to encrypt a message $M \in {0,1}^l$ only decryptable at the epoch $p$ will perform the following:

> 1. Compute $\mathbb{G_{id}}= e(P, Q_{id}) = e(P,H_1(p))$, the "round public key" 

Reviewer: We're using a public key on G2 now, so it's the converse for $P\in \mathbb{G_2}$ the group public key.
```suggestion
1. Compute $\mathbb{G_{id}}= e(Q_{id},P) = e(H_1(p),P)$, the "round public key" 
```

> - (this can be pre-computed per epoch, it's the same for everyone)

Reviewer: ```suggestion
    - (this can be pre-computed per epoch, it's the same for everyone)
```

> 

Reviewer: ```suggestion
```

Reviewer: Not sure how it's rendering but usually lists in MD are not separated with new lines as it would mean a new list each time with a single item

2. Choose a random $(\theta \in {0,1}^l)$, "the mask"

> 

Reviewer: ```suggestion
```

3. Set $r = H_3(\theta, M)$ where $H_3:{0,1}^* \to F_q$ is a secure hash function, "the ephemeral secret key"

> 

Reviewer: ```suggestion
```

4. Output the ciphertext $C = {U, V, W}$ where:

>     $U = rG_1,$ "the ephemeral public key"

Reviewer: ```suggestion
    $U = rG_2\in \mathbb{G_2}$, "the ephemeral public key"
```

    $V = \theta \oplus H_2(rG_{id})$, "the mask commitment"

    $W = M \oplus H_4(\theta)$, "the one-time pad"

### Decryption

> A client that wishes to decrypt a ciphertext $V$ takes the associated signature of epoch $p: \pi_p$ and performs the following:

Reviewer: ```suggestion
A client that wishes to decrypt a ciphertext $C$ takes the associated signature of epoch $p: \pi_p \in \mathbb{G_1}$ and performs the following:
```

> 1. Compute $\theta = V \oplus H_2(e(U, \pi_p))$

Reviewer: ```suggestion
1. Compute $\theta = V \oplus H_2(e(\pi_p, U))$
```

> 

Reviewer: ```suggestion
```

2. Compute $M = W \oplus H_4(\theta)$

> 

Reviewer: ```suggestion
```

3. Set $r = H_3(\theta, M)$

> 

Reviewer: ```suggestion
```

> 4. Test that $U = rG_1 \to$ if not, reject

Reviewer: ```suggestion
4. Test that $U = rG_1 $ if not, reject
```

> 

Reviewer: ```suggestion
```

5. $M$ is the corresponding plaintext

## Libraries available for timelock encryption

Listed below are various libraries and clients for using timelock encryption. Details on how to install and use them are available at the links provided.

- [tlock](https://github.com/drand/tlock)

tlock is a go CLI and library developed by [Ardan Labs](https://www.ardanlabs.com/) and maintained by the drand team for performing timelock encryption and decryption

- [tlock-js](https://github.com/drand/tlock-js)

tlock-js is a javascript library developed and maintained by the drand team for performing timelock encryption and decryption written in Typescript

- [dee](https://github.com/thibmeu/drand-rs)

dee is a rust CLI developed and maintained by Thibault Meunier from cloudflare. It can perform timelock encryption and decryption, as well as retrieve beacons from any of the drand networks.

- [timevault](https://timelock.drand.love)

timevault is a web app for timelock encrypting and decrypting vulnerability reports for responsible disclosure as well as arbitrary text. It is developed and maintained by the drand team.

- tlock-fvm

> Timelock encryption and decryption will be available on the [Filecoin Virtual Machine](https://fvm.bilecoin.io/) in the second half of 2023. Stay tuned to our blog for updates!

Author: Yes, the extra newline after them does that - I corrected an error with the fvm one because it didn't have a newline

> Timelock encryption and decryption will be available on the [Filecoin Virtual Machine](https://fvm.bilecoin.io/) in the second half of 2023. Stay tuned to our blog for updates!

Author: Similarly for the other list items it seemed to work fine with newlines

> Timelock encryption and decryption will be available on the [Filecoin Virtual Machine](https://fvm.bilecoin.io/) in the second half of 2023. Stay tuned to our blog for updates!

Author: You can check a preview of it in the fleek/build action in github actions btw
